pocs/camera/camera.py

```diff
@@ -155,14 +172,101 @@ def CCD_cooling_power(self):
+                describing the observation
+            headers (dict): Header data to be saved along with the file.
+            **kwargs (dict): Optional keyword arguments (`exp_time`, dark)
+
+        Returns:
+            threading.Event: An event to be set when the image is done processing
+        """
+        # To be used for marking when exposure is complete (see `process_exposure`)
+        camera_event = Event()
+
+        exp_time, file_path, image_id, metadata = self._setup_observation(observation,
+                                                                          headers,
+                                                                          filename,
+                                                                          *args,
+                                                                          **kwargs)
+
+        exposure_event = self.take_exposure(seconds=exp_time, filename=file_path, *args, **kwargs)
+
+        # Add most recent exposure to list
+        if self.is_primary:
+            observation.exposure_list[image_id] = file_path
+
+        # Process the exposure once readout is complete
+        t = Thread(target=self.process_exposure, args=(metadata, camera_event, exposure_event))
+        t.name = '{}Thread'.format(self.name)
+        t.start()
+
+        return camera_event
 
     def take_exposure(self, *args, **kwargs):
         raise NotImplementedError
 
-    def process_exposure(self, *args, **kwargs):
-        raise NotImplementedError
+    def process_exposure(self, info, signal_event, exposure_event=None):
+        """
+        Processes the exposure.
+
+        If the camera is a primary camera, extract the jpeg image and save metadata to mongo
+        `current` collection. Saves metadata to mongo `observations` collection for all images.
+
+        Args:
+            info (dict): Header metadata saved for the image
+            signal_event (threading.Event): An event that is set signifying that the
+                camera is done with this exposure
+            exposure_event (threading.Event, optional): An event that should be set
+                when the exposure is complete, triggering the processing.
+        """
+        # If passed an Event that signals the end of the exposure wait for it to be set
+        if exposure_event is not None:
+            exposure_event.wait()
+
+        image_id = info['image_id']
+        seq_id = info['sequence_id']
+        file_path = info['file_path']
+        self.logger.debug("Processing {}".format(image_id))
+
+        try:
+            self.logger.debug("Extracting pretty image")
+            img_utils.make_pretty_image(file_path,
+                                        title=info['field_name'],
+                                        primary=info['is_primary'])
+        except Exception as e:
+            self.logger.warning('Problem with extracting pretty image: {}'.format(e))
+
+        self._process_fits(file_path, info)
```

Comment:
For the gphoto2 class, this method puts the name of the fits file into `info`, replacing the cr2 file. However down below on line 259 the `file_path` (still `.cr2`) is passed into `fpack` as opposed to `info['file_path']`. Either need to return the file here so `file_path = self._process_fits(file_path, info)` or to duplicate line 238 below this line. Or change 259 to `...fpack(info['file_path']`.

That's the only change I had to make and everything else is looking good!

Reply:
Ah, OK. I opted to have `_process_fits()` return the FITS file path, and changed this line to `file_path = self._process_fits(file_path, info)`.

Reply:
Yes I think that makes the most sense. Will test shortly.